### PR TITLE
preload val_map from orders

### LINKED ISSUE #2084: Facet category order with missing data
This seems like a bug: https://community.plot.ly/t/python-plotly-express-facet-col-facet-row-order-categories/33693

### PR DESCRIPTION
**Pull request by nicolaskruchten**
Closes #2084 by populating `val_map` in a group-independent way. @emmanuelle  thoughts on the approach?

### CODE REVIEW

File: packages/python/plotly/plotly/express/_core.py [hunk 1]
```diff
@@ -1162,7 +1164,7 @@ def get_orderings(args, grouper, grouped):
                 key=lambda g: orders[col].index(g[i]) if g[i] in orders[col] else -1,
             )
 
-    return orders, group_names
+    return orders, group_names, group_values
```
Comment (emmanuelle) [inline packages/python/plotly/plotly/express/_core.py]: what is the difference between `orders` and `group_values`? In the examples I tried (including with facets and user-specified category orders) they were the same (the keys were not in the same order, but keys of dicts are not ordered I think?)

Comment (nicolaskruchten) [reply packages/python/plotly/plotly/express/_core.py]: `orders` includes all values given by the user even if they don't show up in the data

Comment (nicolaskruchten) [reply packages/python/plotly/plotly/express/_core.py]: whereas `group_values` has keys only per-group and has only values that show up in the data, so it's more specific. I'm pretty sure we need both, or at least I'm not comfortable trimming down `orders` at the moment :)

Comment (emmanuelle) [reply packages/python/plotly/plotly/express/_core.py]: Thanks. Would you mind adding a comment? The number of variables exceeds by working memory :-)

Comment (nicolaskruchten) [reply packages/python/plotly/plotly/express/_core.py]: pushed!

Comment (emmanuelle) [reply packages/python/plotly/plotly/express/_core.py]: thanks

File: packages/python/plotly/plotly/tests/test_core/test_px/test_px.py [hunk 2]
```diff
@@ -182,3 +182,43 @@ def test_px_templates():
...
+                facet_col="day",
+                color="time",
+                symbol="day",
+                symbol_sequence=symbol_sequence,
+                color_discrete_sequence=color_sequence,
+                category_orders=dict(day=days, time=times),
+            )
+
+            for col in range(len(days)):
+                for trace in fig.select_traces(col=col + 1):
```
Comment (emmanuelle) [inline packages/python/plotly/plotly/tests/test_core/test_px/test_px.py]: oh great I didn't know `fig.select_traces` :-). Not sure it's documented anywhere.

File: packages/python/plotly/plotly/tests/test_core/test_px/test_px.py [hunk 3]
```diff
@@ -182,3 +182,43 @@ def test_px_templates():
     assert fig.layout.xaxis3.showgrid is None
     assert fig.layout.yaxis2.showgrid
     assert fig.layout.yaxis3.showgrid
+
+
+def test_orthogonal_orderings():
```
Comment (emmanuelle) [inline packages/python/plotly/plotly/tests/test_core/test_px/test_px.py]: great test!

Comment (emmanuelle) [inline packages/python/plotly/plotly/tests/test_core/test_px/test_px.py]: Is it worth testing that adding in `category_orders` a value not present in the data does not break things?

**Comment by emmanuelle** [comment]
The approach (defining col and row numbers in advance) looks sound to me, and the test is very useful. :dancer:

File: packages/python/plotly/plotly/tests/test_core/test_px/test_px.py (hunk 3 above)
Comment (nicolaskruchten) [reply packages/python/plotly/plotly/tests/test_core/test_px/test_px.py]: yeah I tried but it's a harder test to write and much slower :)

Comment (nicolaskruchten) [reply packages/python/plotly/plotly/tests/test_core/test_px/test_px.py]: OK I think I have a way to spot-check a few options

**Pull request merged**